Thriveopedia and Fossilisation

Oliveriver opened the pull request:
**Brief Description of What This PR Does**

Adds a rudimentary Thriveopedia and the ability to fossilise species.

The Thriveopedia currently contains:

* A home page
* A museum containing all fossilised species
* A current world page displaying world generation settings for the current world
* The patch map for the current world
* The evolutionary tree for the current world

<details>
<summary>Open for screenshots</summary>

![2022-10-12_22 43 27 4666](https://user-images.githubusercontent.com/7430433/195453972-794bda9a-a564-4fc4-9cf6-319e939201c1.png)

![2022-10-12_22 43 33 2833](https://user-images.githubusercontent.com/7430433/195453976-7f0b83d4-053f-4907-aec1-ba701ba94b34.png)

![2022-10-12_22 43 50 8865](https://user-images.githubusercontent.com/7430433/195453982-414a9114-f94f-4db7-a0b9-176a92597ddb.png)

![2022-10-12_22 43 54 9025](https://user-images.githubusercontent.com/7430433/195453986-0830fe46-f566-4eff-ac59-0dce11239d86.png)

![2022-10-12_22 43 58 5363](https://user-images.githubusercontent.com/7430433/195453987-579c922d-a3c8-4513-885a-7d3b5f00c29a.png)

![2022-10-12_22 44 01 8680](https://user-images.githubusercontent.com/7430433/195453989-910c5efd-3c69-41ad-a5a3-0b728bba5445.png)

![2022-10-12_22 44 27 3876](https://user-images.githubusercontent.com/7430433/195453993-a870f330-479b-4ef4-b372-c2866f082fe5.png)
</details>

**Related Issues**

Closes https://github.com/Revolutionary-Games/Thrive/issues/289

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

Oliveriver commented:
Changing this to a proper PR since I've finally added all the functionality I wanted to add. That's not to say this is done - there are still several bugs to fix detailed below and a whole load of cleanup/linter appeasement required - but it's at least in a state where people can reliably test it. I hope.

Remaining known bugs I'll fix in this PR:

* Gaps between panels in the Thriveopedia aren't consistent. (Easy)
* Fossilisation requires a new icon, as does expanding/collapsing the page tree. (Easy)
* If a species is fossilised in the current game, the museum in the pause menu doesn't update. (Easy)
* Various minor margin width changes. (Easy, but tedious)
* Museum card border highlights stick around after hovering. (Easy, probably)
* Saves are currently broken as the species history doesn't serialize properly. (Medium)
* The number of columns in the museum doesn't update to reflect the expansion/contraction of the page tree. (Medium)
* The evolutionary tree sometimes fails to build for no apparent reason. I suspect this is due to a species going extinct. (Hard)
* A very strange bug sometimes occurs when loading a fossilised AI-created species in freebuild. The ATP production bar and tooltip break due to missing data, causing a crash. Any leads here would be welcome. (Hard)

Remaining known bugs and enhancements I'm proposing leaving out of this PR as out of scope:

* The evolutionary tree fails to build in freebuild games.
* The evolutionary tree fails to build when loading a save made before this PR.
* The player species node in the evolutionary tree always updates to reflect the present build. Ideally, there should be one node per generation showing its progression across time.
* We could add Primum Thrivium as an already existing fossilised species to demonstrate the museum's function.
* If the player moves patches during an editor session, the patch map in the Thriveopedia doesn't update until they leave the editor. Since the Thriveopedia gets this information from the GameProperties object and this isn't updated until the player exits the editor, I don't have any idea how to fix this. It is very minor though.
* Add fossilisation buttons to species in the evolutionary tree and editor.

buckly90 commented:
Great work! The Thriveopedia is looking sleek and clean, not that there is much to clutter it.

**Issues Found:**
1. Hitting the home button brings you to the home page as intended, but the page list still highlights the previous page as if it is selected. You cannot return to the previous page until you select another, and then select it again when this happens.
2. The close button doesn't seem to work at all if you access the pedia from the main menu. It works in-game though!

**Suggestions**:
1. If possible, replacing the fossilize icon over a species with a different one when they have already been fossilized would be nice. Just changing the color could work too.

Oliveriver commented:
> 1. Hitting the home button brings you to the home page as intended, but the page list still highlights the previous page as if it is selected. You cannot return to the previous page until you select another, and then select it again when this happens.

Good spot. I'll see if I can fix that.

> 2. The close button doesn't seem to work at all if you access the pedia from the main menu. It works in-game though!

I'm pretty sure this is the intermittent Godot connection issue that also sometimes happens with other screens in the main menu. I can't reproduce it at the moment.

> 1. If possible, replacing the fossilize icon over a species with a different one when they have already been fossilized would be nice. Just changing the color could work too.

Good idea.

Oliveriver commented:
Update: I've now fixed all the bugs from my reply above and addressed the issues Buckly found, _except_ the tree breaking due to a species going extinct. Saves from both this branch and master appear to be loading correctly.

Tomorrow I'll tidy everything up and try to fix the broken tree issue if I have time. Then this should finally be ready for review.

Oliveriver commented:
Right. I believe this is finally ready for review. I found and fixed a few more bugs in the page tree while cleaning things up, and according to all my testing since, everything works, aside from two things I've decided I need help with.

* The evolutionary tree in the Thriveopedia breaks whenever a species goes extinct from the world. I had a quick look into this but couldn't find anything obvious. There will be people who know more than me about this though.

* Not a bug per se, but definitely a suboptimal way of doing things. I had real trouble saving and loading the RunResults needed to build the evolutionary tree. It turns out JSON can't properly serialize dictionaries whose keys are custom types, and there's a dictionary with Patch objects as keys I needed to save. I found an existing dictionary converter in the Thrive code, but that seemingly only works for static types (compounds, enzymes) and not types such as patches, which are generated dynamically and can change. My eventual solution was to convert the dictionary to and from a list of key-value pairs on saving and loading in RunResults itself. Why not just use the list of key-value pairs everywhere instead, you ask? I tried that and ended up with so many changes in other files that I ran a serious risk of breaking something completely unrelated to this PR. In summary, if someone can write a dictionary converter for patches, I'd be very grateful.

84634E1A607A commented on src/general/GameWorld.cs:
```diff
@@ -47,10 +51,21 @@ public class GameWorld : ISaveLoadable
         WorldSettings = settings;
-        PlayerSpecies = CreatePlayerSpecies();
+
+        if (startingSpecies == null)
```
Good point. Can apply to Auto-evo Exploring Tool as well. - New Issue?

hhyyrylainen replied:
Well isn't the case in the exploring tool that there is *always* the player species present? Which we determined to be actually required to make the "enter game world" button to work sensibly there as there needs to be a player species to control.

84634E1A607A replied:
Well my first design was that you can choose any species and enter the game as it.

hhyyrylainen replied:
So it does make sense in general, then to fix the case with no starting species.

84634E1A607A commented:
![image](https://user-images.githubusercontent.com/60734649/197662364-fb376f1e-23d3-4197-af54-1b945456614e.png)
Yet another bug I've created.

84634E1A607A commented on src/gui_common/thriveopedia/ThriveopediaEvolutionaryTreePage.cs:
```diff
@@ -0,0 +1,158 @@
+            foreach (var generation in CurrentGame.GameWorld.GenerationHistory)
+            {
+                var record = generation.Value;
+                evolutionaryTree.UpdateEvolutionaryTreeWithRunResults(
```
It looks like my previous comment has something to do with this. It looks like this makes Tree believe that a species is split from another of the same generation.

hhyyrylainen commented:
> It turns out JSON can't properly serialize dictionaries whose keys are custom types, and there's a dictionary with Patch objects as keys I needed to save. I found an existing dictionary converter in the Thrive code, but that seemingly only works for static types (compounds, enzymes) and not types such as patches, which are generated dynamically and can change.

`Species` class is used as dictionary keys and setup to be auto string serialized and converted back. It's been a while since I've looked at that code but getting that same system to work for `Patch` class should be easy enough. If you can't figure it out I can use some of my time to re-familiarize myself with this system and figure out the parts you need to set. It might be as simple as adding `[TypeConverter(typeof(ThriveTypeConverter))]` to the `Patch` class.

hhyyrylainen commented on Thrive.csproj:
```diff
@@ -140,6 +140,13 @@
     <Compile Include="src\gui_common\DismissibleNotice.cs" />
     <Compile Include="src\gui_common\FocusFlowDynamicChildrenHelper.cs" />
     <Compile Include="src\gui_common\PhotographablePreview.cs" />
+    <Compile Include="src\gui_common\thriveopedia\Thriveopedia.cs" />
```
I think as the thriveopedia is a pretty big feature by itself (maybe even as big as some editors) that it would kind of make sense for it to be not under gui_common folder, instead being directly under "src", as it's not really a common GUI component building block. And the fossilization folder could also then put under the thriveopedia folder to group those interrelated functionalities together.

hhyyrylainen commented on Thrive.csproj:
```diff
@@ -466,6 +477,8 @@
     <Compile Include="src\modding\NewModGUI.cs" />
     <Compile Include="src\modding\WorkshopData.cs" />
     <Compile Include="src\saving\FileHelpers.cs" />
+    <Compile Include="src\saving\FossilisedSpecies.cs" />
+    <Compile Include="src\saving\FossilisedSpeciesInformation.cs" />
```
Specific saveable types shouldn't be in the saving folder but once again in the thriveopedia subfolder I suggested above.

hhyyrylainen commented on src/general/WorldGenerationSettings.cs:
```diff
@@ -32,14 +34,22 @@ public WorldGenerationSettings()
 
     public enum LifeOrigin
     {
+        [Description("LIFE_ORIGIN_VENTS")]
```
What are these? I don't remember seeing these before...

Oliveriver replied:
An approach for translating enum values (see the new methods at the end of this file). There's an example in `SaveInformation` already.

hhyyrylainen replied:
I see there's already `Description` in the translation extraction script so I suppose this is fine, though this is one more thing for new people to understand about the translation system...

hhyyrylainen commented on src/auto-evo/EvolutionaryTree.cs:
```diff
@@ -167,6 +167,17 @@ public void Init(Species luca)
+        generationTimes.Clear();
+        speciesNodes.Clear();
+
+        foreach (Node child in tree.GetChildren())
```
There's a helper method already in the game to free all children.

hhyyrylainen commented on src/auto-evo/GenerationRecord.cs:
```diff
@@ -0,0 +1,37 @@
+        ///   Auto-Evo results for this generation.
+        /// </summary>
+        [JsonProperty]
+        public RunResults AutoEvoResult = null!;
```
In this case as this is a normal C# class I'd prefer to have a constructor that takes in the AutoEvoResults and AllSpecies, that way the code doesn't need null suppression and is much more likely to be used correctly without causing null bugs.

You can see the Patch and region and the map classes on how I handled the JSON constructors etc. there.

Oliveriver commented:
> `Species` class is used as dictionary keys and setup to be auto string serialized and converted back. It's been a while since I've looked at that code but getting that same system to work for `Patch` class should be easy enough. If you can't figure it out I can use some of my time to re-familiarize myself with this system and figure out the parts you need to set. It might be as simple as adding `[TypeConverter(typeof(ThriveTypeConverter))]` to the `Patch` class.

I definitely tried adding `[TypeConverter(typeof(ThriveTypeConverter))]` to `Patch` at some point and it didn't work. Can't remember exactly what went wrong though and I may have made a mistake somewhere, so it's worth another try.

hhyyrylainen commented on src/general/WorldGenerationSettings.cs:
```diff
@@ -124,4 +134,48 @@ public override string ToString()
+    public string GetTranslatedPlanetString()
+    {
+        return string.Format(CultureInfo.CurrentCulture, TranslationServer.Translate("PLANET_DETAILS_STRING"),
+            TranslationServer.Translate(MapType.GetAttribute<DescriptionAttribute>()?.Description),
```
Doesn't this pretty much completely fail to fulfill the purpose of this method if the getattribute fails? I think a cleaner approach would be to have translation helper take in a generic enum value (if possible) and if not possible then here using an attribute get overload that throws on failure to find would be at least a bit better.

Oliveriver replied:
I'm struggling to parse exactly what you mean here. I agree it would be better to have an approach that throws if no description is found but I can't figure out what your specific suggestion is.

hhyyrylainen replied:
Modify `GetAttribute` to return `T` and throw if it can't find the value.

Actually it looks like the method exists already before this PR so it might be a bit tricky to modify it depending on where it is used elsewhere, but I'd really like it to look like this:
```
    public static T GetAttribute<T>(this Enum e)
        where T : Attribute
    {
        return GetAttributes<T>(e).First();
    }
```

Oliveriver replied:
The only other place this is used throws if it returns null anyway, so this should be fine.

buckly90 commented:
I tried my hardest, but I couldn't seem to get anything to break. Everything functioned perfectly well, superb job!

However I can't help but wonder how fossilization would handle modded entities, such as custom parts. If you fossilize a modded creature, and then remove the mod, would that crash the game when opening the thriveopedia? This could potentially be pretty bad. Unfortunately I don't know of any mods that add parts so I can't really test it.

hhyyrylainen commented:
I'm not really sure if we should do anything about that. There's about five billion things a mod could do that make thrive crash. I guess that when a mod is uninstalled there's just a few ways to still cause crashes, and there's an issue for save compatibility warning with different mods: https://github.com/Revolutionary-Games/Thrive/issues/3283 Though I'll leave it up to @Oliveriver whether you want to add some guard stuff to show an error when loading incompatible fossilized species instructing the user which folder to delete or reinstall mods.

But on the other hand Thrive version updates could also potentially make fossilized species incompatible, though for that it'd probably be easier to just do the same thing as with saves and put a version number into the fossilized species file and refuse to load non-matching ones.

hhyyrylainen commented:
> I could add a try-catch block in museum card rendering and leave it at that.

I haven't had time to read all of the code, but unless the catch is where the data is tried to be deserialized or where the species is applied, I don't think a catch is going to actually really do anything...

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaPatchMapPage.cs:
```diff
@@ -0,0 +1,112 @@
+/// </summary>
+/// <remarks>
+///   <para>
+///     Note a lot of this functionality is duplicated from PatchMapEditorComponent.
```
For cleaner code would it be possible to create a new class that is used by composition by both this and PatchMapEditorComponent so that most of the tricky logic could be included there (I also saw that this at least duplicates the seed display scene part, so maybe that new class should be a higher level wrapper scene on the underlying map component to provide the common functionality)?

Oliveriver replied:
I've just spent several hours trying to get this to work and gave up. There's so much interdependence in the editor classes that I couldn't separate neatly.

I got it working but saves were broken, and any attempt to fix it created horrid spaghetti code. Honestly making such a major change to the editor would introduce so much more risk for bugs and extra review requirements that I doubt it's a good idea for me to do it here.

If you want to fix it go ahead, but I don't think I can.

hhyyrylainen replied:
Sounds like a pain. I feel like someone has finally experienced the pain I had with trying to setup the later stage scenes when I needed to refactor almost everything in the game and HUD scenes...

We can open a follow up issue for this, though until that is done people need to be careful not to accidentally make changes in only one place the map is shown.

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaHomePage.cs:
```diff
@@ -0,0 +1,18 @@
+public class ThriveopediaHomePage : ThriveopediaPage
+{
+    public override string PageName => "Home";
+    public override string TranslatedPageName => TranslationServer.Translate("HOME_PAGE");
```
I think the thriveopedia pages should have "THRIVEOPEDIA_" prefix, and maybe "_TITLE" suffix for all of the page names for super clarity.

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaHomePage.cs:
```diff
@@ -0,0 +1,18 @@
+/// </summary>
+public class ThriveopediaHomePage : ThriveopediaPage
+{
+    public override string PageName => "Home";
```
This is like an internal name, right? If so this should be lower camelCase to be consistent with the rest of the game.

Oliveriver replied:
It's used as part of the name for the scene loaded for each page, so needs to be PascalCase there.

hhyyrylainen replied:
I did eventually get that far in the code. Saying that explicitly in the base abstract class xml documentation would be nice.

hhyyrylainen commented on src/general/StageHUDBase.cs:
```diff
@@ -477,6 +490,9 @@ public override void _Ready()
         sunlight = SimulationParameters.Instance.GetCompound("sunlight");
         temperature = SimulationParameters.Instance.GetCompound("temperature");
 
+        fossilisationButtonLayer = GetNode<Control>(FossilisationButtonLayerPath);
```
Don't these crash the macroscopic prototype now?

hhyyrylainen commented on src/general/StageHUDBase.cs:
```diff
@@ -716,6 +736,52 @@ public void UpdateEnvironmentalBars(BiomeConditions biome)
+    /// <summary>
+    ///   Creates and displays a fossilisation button above each on-screen microbe.
+    /// </summary>
+    public void ShowFossilisationButtons()
```
This should *not* be in the base hud, instead should be abstract here and concrete classes should have this.

hhyyrylainen commented on src/general/StageHUDBase.cs:
```diff
@@ -716,6 +736,52 @@ public void UpdateEnvironmentalBars(BiomeConditions biome)
+    /// <param name="button">The button attached to the microbe to fossilise</param>
+    public void ShowFossilisationDialog(FossilisationButton button)
+    {
+        fossilisationDialog.SelectedSpecies = button.AttachedMicrobe.Species;
```
I think the fossilization dialog should use the base "Species" instead of referring to microbes as this is in the base HUD which means this needs to work for all future stages as well. I'm fine with if you specifically just put a "Not implemented for later stages" message in the fossilize dialog when opened for a non-microbe species.

hhyyrylainen commented on src/general/StageHUDBase.cs:
```diff
@@ -1390,4 +1456,12 @@ private void UpdatePausePrompt()
+
+    private void UpdateFossilisationButtons()
+    {
+        foreach (FossilisationButton button in fossilisationButtonLayer.GetChildren())
```
I'd prefer to have an early return in this method if the fossilizationButtonLayer is not visible.

hhyyrylainen commented on src/general/StageHUDBase.cs:
```diff
@@ -716,6 +736,52 @@ public void UpdateEnvironmentalBars(BiomeConditions biome)
+                nameof(ShowFossilisationDialog));
+
+            // Display a faded button with a different hint if the species has been fossilised.
+            var alreadyFossilised = FossilisedSpecies.CreateListOfSaves().Contains(
```
Calling CreateListOfSaves in this loop sounds like a really bad idea to me... would be better to build that list outside this loop.

hhyyrylainen commented on src/general/WorldGenerationSettings.cs:
```diff
@@ -124,4 +134,48 @@ public override string ToString()
+            PlayerDeathPopulationPenalty,
+            TranslationServer.Translate("PERCENTAGE_VALUE").FormatSafe(Math.Round(GlucoseDecay * 100, 1)),
+            OsmoregulationMultiplier,
+            TranslationHelper.TranslateBool(FreeGlucoseCloud),
+            TranslationHelper.TranslateBool(PassiveGainOfReproductionCompounds),
+            TranslationHelper.TranslateBool(LimitReproductionCompoundUseSpeed));
```
I see now where the TranslateBool is used, so I'd say rename this to "TranslateFeatureFlag" and then the keys would be "FEATURE_ENABLED" and "FEATURE_DISABLED" and maybe even the English text could be, "enabled" and "disabled" to make the output more natural sounding to non-programmers.

hhyyrylainen commented on src/general/WorldGenerationSettings.cs:
```diff
@@ -124,4 +134,48 @@ public override string ToString()
+    /// </summary>
+    public string GetTranslatedMiscString()
+    {
+        return string.Format(CultureInfo.CurrentCulture, TranslationServer.Translate("MISC_DETAILS_STRING"),
```
```suggestion
        return string.Format(CultureInfo.CurrentCulture, TranslationServer.Translate("WORLD_MISC_DETAILS_STRING"),
```

hhyyrylainen commented on src/thriveopedia/Thriveopedia.cs:
```diff
@@ -0,0 +1,338 @@
+            selectedPage = value;
+            selectedPage.Show();
+
+            // The home page is always the first in the history, so ignore it
+            backButton.Disabled = pageHistory.Count <= 1;
+            forwardButton.Disabled = pageFuture.Count == 0;
+
+            pageTitle.Text = SelectedPage.TranslatedPageName;
+            SelectInTreeWithoutEvent(selectedPage);
```
These seem like things that also happen when next / previous is pressed. So maybe this should be extracted as a separate method?

hhyyrylainen commented on src/thriveopedia/Thriveopedia.cs:
```diff
@@ -0,0 +1,338 @@
+            // Add all pages associated with a game in progress
+            if (currentGame != null)
+            {
+                AddPage("CurrentWorld");
```
Related to my earlier comment about "Home" capitalization, is it actually a node name? If so then it is fine as is, and well to avoid my future confusion you could add a comment on that base property.

hhyyrylainen commented on src/thriveopedia/Thriveopedia.cs:
```diff
@@ -0,0 +1,338 @@
+    /// </summary>
+    /// <param name="name">The name of the desired page</param>
+    /// <returns>The Thriveopedia page with the given name</returns>
+    private ThriveopediaPage? GetPage(string? name)
```
Why can the name to look for be null?

Oliveriver replied:
If a page is created with no parent, this method gets called with null and returns null. I suppose when you put it like that it's better to avoid this path and check for null earlier.

hhyyrylainen commented on src/thriveopedia/Thriveopedia.cs:
```diff
@@ -0,0 +1,338 @@
+    {
+        // Avoid adding duplicate pages
+        if (allPages.Keys.Any(p => p.PageName == name))
+            return;
```
Shouldn't it be a serious error to try to add a duplicate page or?

hhyyrylainen commented on src/thriveopedia/Thriveopedia.cs:
```diff
@@ -0,0 +1,338 @@
+        allPages.Add(homePage, CreateTreeItem(homePage, null));
+
+        // Add all pages not associated with a game in progress
+        AddPage("Museum");
```
I assume this doesn't take too much time because otherwise we should think about only initializing the Thriveopedia when it becomes visible.

Oliveriver replied:
Right now it doesn't, since for the museum in particular the cards are only created when it becomes visible.

It may take longer if the list of pages grows a lot bigger.

hhyyrylainen commented on src/thriveopedia/Thriveopedia.cs:
```diff
@@ -0,0 +1,338 @@
+
+    private void Exit()
+    {
+        // Reset to the home page for the next session
```
When opened in-game it would be super useful if the history was kept between opening and closing the thriveopedia.

Oliveriver replied:
I think I tried this as a fix for a bug I found with the page tree, found a better fix, and accidentally left this in. So yeah, I'll remove it.

hhyyrylainen commented on src/thriveopedia/ThriveopediaPage.cs:
```diff
@@ -0,0 +1,67 @@
+    {
+        base._Ready();
+
+        backgroundPanel = GetNode<PanelContainer>(".");
```
Isn't background panel then just "this"? So on line 54 it should just be "this.AddStyle..."?

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisationButton.cs:
```diff
@@ -0,0 +1,61 @@
+    /// </summary>
+    public void UpdatePosition()
+    {
+        RectGlobalPosition = camera.UnprojectPosition(AttachedMicrobe.GlobalTransform.origin);
```
I think this should have a guard here against the camera no longer being current similarly to how DebugOverlays does it.

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisationButton.cs:
```diff
@@ -0,0 +1,61 @@
+    /// <summary>
+    ///   The microbe this button is attached to.
+    /// </summary>
+    public Microbe AttachedMicrobe = null!;
```
From looking at things, it should be possible to change this to `Spatial` as this doesn't seem to need to know anything about the species itself.

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisationDialog.cs:
```diff
@@ -0,0 +1,232 @@
+
+    /// <summary>
+    ///   True when one of our (name related) Controls is hovered. This needs to be known to know if a click happened
+    ///   outside the name editing controls, for detecting when the name needs to be validated.
```
Can a lose focus callback from the name box be used instead? As that would work for controller input as well (and keyboard navigation).

Oliveriver replied:
I believe I copied this wholesale from EditorComponentBottomLeftButtons, so if this is changed it's probably best to change that too.

hhyyrylainen replied:
It does seem like it, so that would be good to update as well. I just searched for "OnClickedOffName" and it seems like you didn't attach a callback to this method, so it would seem to be that this cannot currently function correctly.

Oliveriver replied:
Having looked into this, I think I can get away with removing it. The editor blocks any other input until the name is valid which is why this is needed there, but I don't need that here as I'm just disabling the fossilisation button when the name is invalid and the player is free to perform other actions.

hhyyrylainen replied:
That sounds fine.

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisationDialog.cs:
```diff
@@ -0,0 +1,232 @@
+        get => selectedSpecies;
+        set
+        {
+            selectedSpecies = (Species)value.Clone();
```
This clone here is a bit unexpected to me to be here, as basically always you assume a property set actually *sets* the value. Would it be possible to refactor so to keep this as-is but then create the cloned species when required?

Oliveriver replied:
It's occured to me that this begs a design choice I'd implicitly answered and ignored: if a player fossilises a species with a new name, should its name change in the current game too? I added this to stop that happening, but we do have the option of using this as a method for renaming species as people have been asking for.

hhyyrylainen replied:
I'd disallow that. The intended future feature is anyway to attach labels to things the player names, not replace their names entirely.

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisationDialog.cs:
```diff
@@ -0,0 +1,232 @@
+            SelectedSpecies.Generation,
+            SelectedSpecies.Population,
+            SelectedSpecies.Colour.ToHtml(),
+            string.Join("\n  ", SelectedSpecies.Behaviour.Select(b => b.Key + ": " + b.Value)));
```
Wasn't there behaviour value translations in the auto-evo exploring tool?

hhyyrylainen replied:
Or is this code exactly the same as in there? If so this is pretty complex so I'd say this should be made into a helper method somewhere (probably the Species class makes most sense).

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisationDialog.cs:
```diff
@@ -0,0 +1,232 @@
+    {
+        GD.Print("Saving species " + SelectedSpecies.FormattedName);
+
+        if (FossilisedSpecies.CreateListOfSaves()
+            .Any(s => s == SelectedSpecies.FormattedName + Constants.FOSSIL_EXTENSION_WITH_DOT))
```
I'm seeing this logic for the second time so I'd say that FossilizedSpecies should have a method to check by *name* if something is fossilized already (and it could take an optional fossils list if called in a loop for example for more efficiency).

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisedSpecies.cs:
```diff
@@ -0,0 +1,243 @@
+/// </summary>
+public class FossilisedSpecies
+{
+    public const string SAVE_SAVE_JSON = "save.json";
```
```suggestion
    public const string SAVE_FOSSIL_JSON = "fossil.json";
```
I'm going to suggest a bunch of changes to make this more distinct from the normal saves.

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisedSpecies.cs:
```diff
@@ -0,0 +1,243 @@
+    /// <summary>
+    ///   The species to be saved/loaded.
+    /// </summary>
+    public Species Species { get; set; } = null!;
```
I really think you should add a constructor to force people to set this right as in my opinion this use of nullability overwrite is an unnecessary way to get around it.

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisedSpecies.cs:
```diff
@@ -0,0 +1,243 @@
+    ///   Creates a list of existing fossilised species names to prevent unintended overwrites.
+    /// </summary>
+    /// <returns>A list of names of .thrivefossil files in the user's home directory</returns>
+    public static List<string> CreateListOfSaves()
```
```suggestion
    public static List<string> CreateListOfFossils()
```

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisedSpecies.cs:
```diff
@@ -0,0 +1,243 @@
+    /// <summary>
+    ///   Creates a list of existing fossilised species names to prevent unintended overwrites.
+    /// </summary>
+    /// <returns>A list of names of .thrivefossil files in the user's home directory</returns>
```
```suggestion
    /// <returns>A list of names of .thrivefossil files in the user's fossils directory</returns>
```

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisedSpecies.cs:
```diff
@@ -0,0 +1,243 @@
+            directory.ListDirEnd();
+        }
+
+        using var file = new File();
+        result = result.OrderBy(item =>
+            file.GetModifiedTime(Path.Combine(Constants.FOSSILISED_SPECIES_FOLDER, item))).ToList();
```
Do these really need to be sorted by time? Well at least when checking if a fossil exist, this should be skipped then.

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisedSpecies.cs:
```diff
@@ -0,0 +1,243 @@
+    /// </summary>
+    /// <param name="saveName">The name of the .thrivefossil file (including extension)</param>
+    /// <returns>The species saved in the provided file</returns>
+    public static Species LoadSpeciesFromFile(string saveName)
```
```suggestion
    public static Species LoadSpeciesFromFile(string fossilName)
```

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisedSpecies.cs:
```diff
@@ -0,0 +1,243 @@
+            throw new NotImplementedException("Saving non-microbe species is not yet implemented");
+        }
+
+        // For now, save everything as a microbe to aid deserialization
```
Well actually due to the dynamic species saving already, it should be possible to really do without the Type etc. fuss here, but unless tested, I think having the NotImplementedException here is still a good idea.

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisedSpecies.cs:
```diff
@@ -0,0 +1,243 @@
+            Type = FossilisedSpeciesInformation.SpeciesType.Microbe,
+        };
+
+        WriteRawSaveDataToFile(speciesInfo, Species.StringCode, Name + Constants.FOSSIL_EXTENSION_WITH_DOT);
```
Is there a reason why the image is not saved in the fossil. I think it would make this much more helpful format if the screenshot of the thing would also be saved. That'd make the museum load much faster, right?

Oliveriver replied:
The reason is I didn't know how to. Just tried it and it would take a significant refactor of the architecture, so I don't think it's worth it for now. There are other, easier ways to get better performance with the museum (like only loading when the Thriveopedia is opened, not when the museum page is).

hhyyrylainen replied:
Doesn't the fossilization dialog show a preview image of the species? That can be grabbed quite easily I think. I'll give that a go with the other saving fixes I'll try to do today.

hhyyrylainen replied:
I've now done this and about to commit. I had to extend the species preview system a bit to get it to save the Images when needed for writing into the fossilized files.

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisedSpecies.cs:
```diff
@@ -0,0 +1,243 @@
+        WriteRawSaveDataToFile(speciesInfo, Species.StringCode, Name + Constants.FOSSIL_EXTENSION_WITH_DOT);
+    }
+
+    private static void WriteRawSaveDataToFile(FossilisedSpeciesInformation speciesInfo, string saveContent,
```
This should also be renamed, I'll now stop pointing out all the instances of the word "save" in this file...

hhyyrylainen commented on src/thriveopedia/fossilisation/FossilisedSpeciesInformation.cs:
```diff
@@ -0,0 +1,45 @@
+    public Guid ID { get; set; } = Guid.NewGuid();
+
+    /// <summary>
+    ///   The type of this species, e.g. microbe.
```
Now that I think about this, I agree it's a good idea to have the stage here, so that museum can be filtered for example without having to load all of the data for each species.

hhyyrylainen commented on src/thriveopedia/fossilisation/MuseumCard.cs:
```diff
@@ -0,0 +1,58 @@
+    /// <summary>
+    ///   The fossilised species associated with this card.
+    /// </summary>
+    public Species SavedSpecies = null!;
```
This should be a nullable property that on change would also reload the preview. Similarly to a bunch of our GUI nodes which either allow setting a value before or after attaching them to a scene.

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaEvolutionaryTreePage.cs:
```diff
@@ -0,0 +1,158 @@
+/// </summary>
+/// <remarks>
+///   <para>
+///     Note a lot of this functionality is duplicated from AutoEvoExploringTool.
```
Again here, I think we should take steps to avoid code duplication to make sure the exploring tool and thriveopedia don't end up with different implementations of things in the long run.

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaEvolutionaryTreePage.cs:
```diff
@@ -0,0 +1,158 @@
+    [Export]
+    public NodePath HexPreviewPath = null!;
+
+    private readonly List<Dictionary<uint, Species>> speciesHistoryList = new();
+    private VBoxContainer disabledInFreebuild = null!;
```
```suggestion
    private readonly List<Dictionary<uint, Species>> speciesHistoryList = new();

    private VBoxContainer disabledInFreebuild = null!;
```
I prefer blank lines between *completely* different categories of variables for easier visual parsing of what's going on here.

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaEvolutionaryTreePage.cs:
```diff
@@ -0,0 +1,158 @@
+        if (!Visible)
+            return;
+
+        // TODO fix the tree for freebuild?
```
```suggestion
        // TODO: fix the tree for freebuild
```

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaEvolutionaryTreePage.cs:
```diff
@@ -0,0 +1,158 @@
+            return;
+
+        // TODO fix the tree for freebuild?
+        if (CurrentGame!.FreeBuild)
```
I'd say do the longer thing and put a `if(CurrentGame == null) throw new invalidoperationexception` at the top of this method instead of just skip over it like this. I've done that in quite many places in the game.

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaEvolutionaryTreePage.cs:
```diff
@@ -0,0 +1,158 @@
+
+        // Building the tree relies on the existence of a full history of generations stored in the current game. Since
+        // we only started adding these in 0.6.0, it's impossible to build a tree in older saves.
+        // TODO avoid an ugly try/catch block by actually checking the original save version?
```
Doing the TODO here would be a pretty good idea

Oliveriver replied:
Is there a good way to go about getting this information? I had a look into this but found no leads.

hhyyrylainen replied:
Well we would actually need a new field in the save info to preserve this info... so kind of probably not worth it (but overall wouldn't be too hard to add, but kind of the point of the save upgrade system is that old saves work as well as new ones). Maybe it'd be just better then to add proper error printing and showing right info to the player (like my next comment suggests).

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaEvolutionaryTreePage.cs:
```diff
@@ -0,0 +1,158 @@
+        }
+        catch (KeyNotFoundException)
+        {
+            evolutionaryTree.Visible = false;
```
Or at least print a TODO error here about the problem. And I haven't looked at the text but I hope the disabledInFreebuild error message also says that it may be because of an older save?

Oliveriver replied:
Yes, maybe that's the wrong name as it covers both scenarios. I'll update that.

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaEvolutionaryTreePage.cs:
```diff
@@ -0,0 +1,158 @@
+
+    private void InitFirstGeneration()
+    {
+        // TODO fix this so it shows player species progression rather than current state
```
```suggestion
        // TODO: fix this so it shows player species progression rather than current state
```

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaEvolutionaryTreePage.cs:
```diff
@@ -0,0 +1,158 @@
+        // TODO fix this so it shows player species progression rather than current state
+        speciesHistoryList.Add(new Dictionary<uint, Species>
+        {
+            { CurrentGame!.GameWorld.PlayerSpecies.ID, CurrentGame.GameWorld.PlayerSpecies },
```
This nullability suppression is more acceptable to me than the one above.

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaEvolutionaryTreePage.cs:
```diff
@@ -0,0 +1,158 @@
+
+    private void UpdateSpeciesDetail(Species species)
+    {
+        speciesDetailsLabel.ExtendedBbcode = TranslationServer.Translate("SPECIES_DETAIL_TEXT").FormatSafe(
```
This is also probably shared code that'd be easy to put in the microbe class, right?

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaMuseumPage.cs:
```diff
@@ -0,0 +1,178 @@
+
+            foreach (var speciesName in FossilisedSpecies.CreateListOfSaves())
+            {
+                var card = (MuseumCard)GD.Load<PackedScene>(
+                    "res://src/thriveopedia/fossilisation/MuseumCard.tscn").Instance();
```
Packed scene should be loaded once in _Ready

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaMuseumPage.cs:
```diff
@@ -0,0 +1,178 @@
+
+                var savedSpecies = FossilisedSpecies.LoadSpeciesFromFile(speciesName);
+
+                if (savedSpecies is not MicrobeSpecies)
+                {
+                    throw new NotImplementedException("Loading non-microbe species is not yet implemented");
+                }
```
I think this throw is a bit much here, I think the card should just reduce the available functions it has if given another type of species.

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaMuseumPage.cs:
```diff
@@ -0,0 +1,178 @@
+
+    private void UpdateSpeciesDetail(Species species)
+    {
+        speciesDetailsLabel.ExtendedBbcode = TranslationServer.Translate("SPECIES_DETAIL_TEXT").FormatSafe(
```
Now *this* is definitely duplicate code.

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaMuseumPage.cs:
```diff
@@ -0,0 +1,178 @@
+        // If we're opening from a game in progress, warn the player
+        if (CurrentGame != null)
+        {
+            leaveGameConfirmationDialog.DialogText = TranslationServer.Translate("OPEN_IN_FREEBUILD_WARNING");
```
```suggestion
            leaveGameConfirmationDialog.DialogText = TranslationServer.Translate("OPEN_FOSSIL_IN_FREEBUILD_WARNING");
```

hhyyrylainen commented on src/thriveopedia/pages/ThriveopediaMuseumPage.cs:
```diff
@@ -0,0 +1,178 @@
+            return;
+        }
+
+        TransitionToFreebuild();
```
I'd say add the previewSpecies as parameter here, then the nullability suppression in that method is not necessary for cleaner code.

hhyyrylainen reviewed (changes requested):
I don't have any real major objections to any of the stuff here, just a bunch of small problems I noticed. I didn't playtest yet but I suspect this breaks the macroscopic prototype entirely, so that at least should be looked at.

Oliveriver commented:
Addressed the many small things. I'll get to the rest tomorrow.

Oliveriver commented:
Right, addressed a few more comments and also improved the museum performance by loading only when the Thriveopedia is opened. For my own benefit, here's what's left:

* Fix the tree failing to build if a species goes extinct. Again, I ran out of time to look into this.
* Extract a common class for the editor patch map and Thriveopedia patch map. After spending several hours on it going nowhere, I've decided I refuse to do it.
* Extract a common class for the exploring tool evolutionary tree and Thriveopedia evolutionary tree. I suspect this will be similar.
* Save images with fossils. Tried it and it was too complex for my brain after all the above. I think it can be safely delegated to an issue to address in future.
* Simplify the dynamic JSON deserialization of species type, which hhyyrylainen will be looking into.

Nearly there...

hhyyrylainen commented:
I found a bug, the overwrite dialog has the title "Confirm Exit":
![2022-11-04_09 37 52 4056](https://user-images.githubusercontent.com/3796411/199918585-36e2aa82-231d-47fb-a608-1add4e026f7d.png)

hhyyrylainen commented:
Another bug:
```
ERROR: Error calling method from signal 'visibility_changed': 'PanelContainer(ThriveopediaPatchMapPage.cs)::RebuildMap': Method not found..
```
I checked that no random scenes are modified so I'm quite sure this is a mistake in this PR.

hhyyrylainen commented:
Okay from playtesting I noticed just one more slight problem which is that the warning about starting a new game from the museum while in-game just talks about starting a new game, but I think the message should say starting a new game as the selected species, just to be clear what's happening. As I think some players might read starting a new game and not click ok as they are confused and thought that they did something wrong.

hhyyrylainen commented:
One more bug:
EarlyMulticellularEditor the statistics button doesn't work

It seems the callback exists but Godot editor *insists* on removing that, and when I reverted everything and directly debugged that scene, the button still didn't work. So could you look at that? I think it might be configured wrong somehow. All the other prototypes did have their statistics buttons working correctly.

Oliveriver commented:
Fixed the small bugs hhyyrylainen found.

Finally figured out what's going on with the broken evolutionary tree. Even though each of the Species in AllSpecies in GenerationRecord is cloned, the Species in AutoEvoResults in GenerationRecord are not. This means the population for every node on each rebuild was always updating to the latest population, not the population at the time. Consequently, when a species went extinct, _all_ nodes for that species were being marked as extinct and so had no non-extinct parent to connect to, causing the KeyNotFoundException.

I've written a fix and will clean it up tomorrow.

hhyyrylainen commented:
> the Species in AutoEvoResults in GenerationRecord are not

I did notice this, but I thought you specifically did it like that. When creating the AutoEvoResults class I did not intend it to be stored past the current auto-evo cycle, so storing it leading to various problems is not very surprising to me.

Oliveriver commented:
Fixed the tree breaking due to species extinction. It required a change to how generation records are stored in saves so any saves made in this branch before that might be broken. But I've tested various saves made on master and the newest version of this branch and found no issues.

Pending further comments on that fix (and one tiny extra fix I had to make to stop old saves crashing when trying to draw the tree), I'm going to call this done. There are a few things outstanding, but at this point I think they're separate enough to break into new issues.

* Add a helper to switch button order in confirmation dialogs depending on system.
* Change the tree drawing to add a node for the player species every generation so they can see their own progression. I had a go at adding this but thought it would be better scrutinised in a new PR, so I'll make that once this is merged.
* Separate common parts of PatchMapEditorComponent and ThriveopediaPatchMapPage into a new scene/class.
* Separate common parts of AutoEvoExploringTool and ThriveopediaEvolutionaryTreePage into a new scene/class.
* Fix the evolutionary tree in freebuild games and old saves (though trees from old saves will always be incomplete at the very least).
* Update the Thriveopedia patch map while in the editor if the player moves patches.
* Add fossilisation buttons to species displays in the evolutionary tree and editor.
* Add the ability to delete fossilised species.

hhyyrylainen commented:
One more thing we can open as separate issue is having a button to open the fossils folder, as I would guess people would like sharing those. People ask enough about how to find the save files.

Oliveriver commented:
> One more thing we can open as separate issue is having a button to open the fossils folder, as I would guess people would like sharing those. People ask enough about how to find the save files.

Yeah, good point. Could easily fit in the museum page.

I've made those last few changes you suggested, so this is ready to merge if you're happy.

hhyyrylainen commented:
I'll do a one more full read through of the code to check my comments were addressed and I don't spot any more problems, in a few days. I'll ask now if anyone else can playtest or code review this before that.

84634E1A607A commented:
Not sure if this is known: When you fossilize a species, the icon will not change to the fossilized state until you resume the game and pause again.

84634E1A607A commented:
Another bug I've written: player mutations are not taken into account. There are no proper nodes indicaitng that the player has evolved.

![image](https://user-images.githubusercontent.com/60734649/200150586-04f96070-33d4-46ab-ae78-d91312ef769f.png)

Oliveriver commented:
> Edit: which seems to be wanted to be merged into this branch. Is there a reason why that's a separate PR, @Oliveriver ?

I did that so it was easier to review.

I've added the changes which minimise save data size based on the findings on Discord earlier. Essentially, I've done three things:

* Replaced actual species data with species IDs for everything except the main species. Turns out recording the full data for every species a given species split from was adding massive bulk.
* Removed unnecessary fields from the saved objects, like per-patch populations which had many patches stored within them. It may be that we want to add these back in future for new features, but for now they're useless bloat.
* Removed species data for the main species if it didn't change in a given generation. This data is then recreated when building the tree using previous generations.

hhyyrylainen commented:
> Removed unnecessary fields from the saved objects, like per-patch populations which had many patches stored within them. It may be that we want to add these back in future for new features, but for now they're useless bloat.

I think patches where the population exist, would be good to save, and still saves over 80% compared to saving all of them.

Did you resolve the generation record having the AllSpecies property and then differently cloned species also stored through the auto-evo results? I saw you remove the generation record class but not modify the GameWorld, so I didn't see how you made that work...

Oliveriver commented:
I didn't remove GenerationRecord, I moved it to a new folder since there are now a few other record classes. Everything that needs to be cloned is, as far as I can tell, still being cloned properly, and species which don't need to be cloned (those that haven't changed) aren't being saved anymore.

hhyyrylainen commented:
I see. Then I just have one comment that I think the SpeciesRecord should probably be SpeciesRecordLite so that the "full" record is fundamentally the more important version. I thought initially that maybe a base class could be made but that doesn't really seem like it as fundamentally the types are meant for different purposes. Though they seem to share at least 2 fields, right? So an abstract base class could be made to save on the documentation string etc. duplication.

hhyyrylainen commented on src/thriveopedia/Thriveopedia.cs:
```diff
@@ -0,0 +1,359 @@
+    {
+        base._Notification(what);
+
+        if (what == NotificationVisibilityChanged && Visible)
```
Well we don't have translations yet but I think some places like the navigation tree needs to be regenerated on language change, I'll open a follow up issue for this as well.

hhyyrylainen commented on src/general/StageBase.cs:
```diff
@@ -285,6 +285,9 @@ public virtual void OnReturnFromEditor()
             throw new InvalidOperationException("Returning to stage from editor without a game setup");
 
+        // Update the generation history with the newly edited player species
+        GameWorld.GenerationHistory.Last().Value.UpdateSpeciesData(GameWorld.PlayerSpecies);
```
This relies on the dictionary being in-order, which is not 100% guaranteed, so the history should actually be a list or what I'll do myself to fix this right now is to find the max key and use that here.

hhyyrylainen commented on src/auto-evo/records/SpeciesRecordLite.cs:
```diff
@@ -0,0 +1,33 @@
+    {
+        [JsonConstructor]
+        public SpeciesRecordLite(Species? species,
+            long population,
+            uint? mutatedPropertiesID = null,
+            uint? splitFromID = null)
```
"chop" formatting of function arguments is not correct style, I'll fix this.

hhyyrylainen reviewed (approved):
I did some more playtesting and looked through the code again. I did find some things to fix (and one editor made save exit editor crash bug) but I fixed those myself so now I think this is good to go.

hhyyrylainen commented:
Opened the follow up issues. I'll highlight just one here: https://github.com/Revolutionary-Games/Thrive/issues/3854 which I marked as high priority because we've historically needed to fix translations not applying in a bunch of places so that'd be nice to fix before anyone complains about the bug.
